src/kagglehub/packages.py

```diff
@@ -0,0 +1,125 @@
+from collections import deque
+from contextlib import contextmanager
+import contextvars
+from dataclasses import dataclass
+from functools import wraps
+import importlib
+import inspect
+import logging
+import pathlib
+import re
+import sys
+from types import ModuleType
+from typing import Callable, Optional, Type
+
+from kagglehub import registry
+from kagglehub.handle import ResourceHandle, parse_package_handle
+from kagglehub.logger import EXTRA_CONSOLE_BLOCK
+from kagglehub.requirements import VersionedDatasources, read_requirements
+
+class PackageScope:
+    """Captures data about a Kaggle Package. Use as Context Manager to apply scope.
+    
+    When scope is applied, certain `kagglehub` calls will utilize the Package info.
+    Specifically, downloading a datasource without a version specified will check
+    the PackageScope and use the Package's version if it finds a matching entry.
+    `kagglehub.get_package_asset_path` also relies on the current scope.
+    """
+
+    _ctx = contextvars.ContextVar("kagglehub_package_scope", default=None)
+    
+    def __init__(self, package_path: pathlib.Path):
+        self.path: pathlib.Path = package_path
+        self.datasources: VersionedDatasources = read_requirements(package_path / "kagglehub_requirements.yaml")
+
+        self._token_stack = deque()
+
+    def __enter__(self):
+        token = PackageScope._ctx.set(self)
+        self._token_stack.append(token)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        token = self._token_stack.pop()
+        PackageScope._ctx.reset(token)
+
+    @staticmethod
+    def get():
+        """Gets the currently applied PackageScope, or None if none applied."""
+        return PackageScope._ctx.get()
+    
+    def apply_to_module(self, module: ModuleType) -> ModuleType:
+        """Decorates all functions/methods in the module to apply our scope."""
+        def apply(obj: ModuleType | Type):
+            for name, member in inspect.getmembers(obj):
+                # Only decorate things defined within the module.
+                if getattr(member, "__module__", None) != module.__name__:
+                    continue
+                # These denylisted entries cause infinite loops
+                elif inspect.isclass(member) and name not in ["__base__", "__class__"]:
```

Comment:
What is the worst case depth of this? Can we use an explicit stack here?

Reply:
This only recurses on classes defined within the user's module, so you'd have to have a ton of nested classes to hit the limit, but fair point to be defensive and it removes one of the helper functions here, fixed.